jekyll/_posts/2021-07-21-Nim-Efficienct-Expressive-Elecgant-Benchmarking.md

```diff
@@ -0,0 +1,369 @@
+3. Haskell actually does have AST macros using "Template Haskell" and if using the optional LLVM backend along with GHC Haskell primitives can produce very efficient code for even the Extreme Loop Unrolling algorithm, which does not depend on memory management/Garbage Collection since just one fixed `seq`/(unboxed primitive)Array is used.  The Haskell source code is very elegant, but where it might lose out in effectiveness is due to its steep learning curve in using it effectively with its very different functional programming paradigms including Monadic code and non-strictness (laziness).  Nim doesn't really compete directly with Haskell as they are two entirely different programming paradigms.  One disadvantage that Haskell has for this algorithm is that "Template Haskell" code needs to be run from a separate source code file.
+4. Languages competitive as to efficiency and general programming paradigm are the fully compiled languages such as C/C++, Chapel, and Rust, all of which can perform the loop unrolling algorithm at about the speed as the above Nim program.  C/C++ and Chapel lose out as to elegance and somewhat as to effectiveness as they don't have proper macros (C/C++ so-called preprocessor "macros" are really just text substitution templates), and Rust, which does have proper AST macros and pointer operations but loses out to being somewhat more verbose (although it gains in safety, which is not one of Nim's primary goals as it is for Rust).  Rust's AST macros ("procedural macros") also currently need to be compiled from a separate source file just as for Haskell's "Template Haskell", thus if one were to want to use them for this algorithm, the result would not be quite as effective nor as elegant as the Nim solution.  The "macro_rules" types of macro are more like what Nim would call templates, and can be used but require invoking the template for each of the 32 varieties of pattern, so neither as effective nor as elegant.
```

Comment:
Safety is regarded most important in Nim, it's just that Nim predates Rust's borrow checking so we're still playing catch-up on this front. However we also have:

- ranged integers
- static deadlock prevention via Nim's "lock levels"
- protection against cyclic memory leaks

(All of which Rust lacks.)

Reply:
I only mentioned Rust's safety as a counter balance to it currently being slightly slower in this benchmark (due to a slight mis-optimization by LLVM), but I can counter that with Nim's ranged integers (that are checked at compile time so no runtime cost, right?).

Ass to deadlock prevention, I know about it but planned to bring it up in a later article showing how with Arc/Orc and the new `channels` and `threadpool` modules, Nim has the threading story covered.

I have another pending post comparing features such as the cyclic reference memory leak prevention; but didn't mention it in this post because this benchmark doesn't need it, and doesn't even need Arc as there are almost no allocations.

I wonder if these things should just be mentioned very briefly in this article (which is already quite long) with a promise to cover them in more depth in a later article.

You don't have to tell me about what Rust lacks as I just finished reviewing my code for the Hamming Number Sequence Generator, and it was both very difficult to do, is about three times the lines of code as Nim no matter how I try to reduce it, and still has that pesky cyclic reference memory leak.  But I don't want the article to be too negative about the competition and rather just very positive about Nim.

Reply:
It's not my intention to "downplay" Rust, we took much inspiration from it. But Nim cares as much about security as Rust does and is ahead in some aspects (cycles, deadlocks) and behind in others (borrow checking).

Btw, the integer ranges are checked dynamically but could be done statically. And will be so in the (far away) future.

Reply:
Yes, you have "vision" derived from many sources, and I can hardly wait to see most or all of experimental features and other tweaks that can be done by static analysis implemented in future Nim.

There are so many of them and not all of them are ready for "prime time" yet that I don't want to talk about all of them in one article, but plan to break things down in the mentioned series of articles.

Reply:
I think in the interests of not increasing the length of this article beyond what it already is, I'll keep the in depth analysis of "safety" to a later article.  However, I can add a statement to this effect after the comment about the comparison to Rust.  Something like the following:
"""
(although it gains in safety, which is not yet one of Nim's primary goals as it is for Rust; however, Nim's future goals do include "safety" about equivalent to that of Rust, which will be discussed in depth in a later article)
"""